src/hotspot/share/opto/escape.cpp

```diff
@@ -448,29 +458,25 @@ bool ConnectionGraph::compute_escape() {
 }
 
 // Check if it's profitable to reduce the Phi passed as parameter.  Returns true
-// if at least one scalar replaceable allocation participates in the merge and
-// no input to the Phi is nullable.
+// if at least one scalar replaceable allocation participates in the merge.
 bool ConnectionGraph::can_reduce_phi_check_inputs(PhiNode* ophi) const {
-  // Check if there is a scalar replaceable allocate in the Phi
   bool found_sr_allocate = false;
 
   for (uint i = 1; i < ophi->req(); i++) {
-    // Right now we can't restore a "null" pointer during deoptimization
-    const Type* inp_t = _igvn->type(ophi->in(i));
-    if (inp_t == nullptr || inp_t->make_oopptr() == nullptr || inp_t->make_oopptr()->maybe_null()) {
-      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. Input %d is nullable.", ophi->_idx, _invocation, i);)
-      return false;
-    }
-
-    // We are looking for at least one SR object in the merge
     JavaObjectNode* ptn = unique_java_object(ophi->in(i));
     if (ptn != nullptr && ptn->scalar_replaceable()) {
       assert(ptn->ideal_node() != nullptr && ptn->ideal_node()->is_Allocate(), "sanity");
       AllocateNode* alloc = ptn->ideal_node()->as_Allocate();
 
+      // Don't handle arrays.
+      if (alloc->Opcode() != Op_Allocate) {
```

Comment:
I suggest to add assert inside branch to catch the case when we add an other Allocate subclass.

Also `ptn->ideal_node()->is_Allocate()` check at the line 468 is not needed because `as_Allocate()` has such assert check.

Reply:
Thanks. Fixed. Going up in next push.